### PR TITLE
MDEV-24523: Execution of JSON_REPLACE failed on Spider

### PR DESCRIPTION
**Pull request by yongxin-xu**
how to repeat:
CREATE TABLE tbl_a (
  i INT,
  j JSON,
  PRIMARY KEY(i)
) ENGINE=SPIDER, PARTITION BY ...

INSERT INTO tbl_a VALUES (1, '{ "a": 1, "b": [2, 3]}');
UPDATE tbl_a SET j = JSON_REPLACE(j, '$.a', 10, '$.c', '[1, 2]');

### CODE REVIEW

File: storage/spider/mysql-test/spider/bugfix/r/mdev_24523.result
Comment (yongxin-xu) [inline storage/spider/mysql-test/spider/bugfix/r/mdev_24523.result:15]: @nayuta-yanagisawa Do I need to change the base branch to 10.3? If needed, I will also add the `disable-ps-protocol` and remove `log_output` like my previous fixes.

File: sql/item_jsonfunc.h
Comment (yongxin-xu) [inline sql/item_jsonfunc.h:528]: I think the `json_update` is a typo here, 'cause there is no function named `json_update` in [MariaDB JSON functions](https://mariadb.com/kb/en/json-functions/).

**Review by yongxin-xu — COMMENTED** [review]
Added some review comments on the code.

File: storage/spider/mysql-test/spider/bugfix/r/mdev_24523.result
Comment (nayuta-yanagisawa) [reply storage/spider/mysql-test/spider/bugfix/r/mdev_24523.result:15]: JSON_XXX function is introduce by https://github.com/MariaDB/server/commit/ebe5ebba165863a134c145b1a3709b70925d5100. The commit is also merged into 10.2? So, the bug might reproduce on 10.2. Could you check if it is reproducible on 10.2? The base branch should be the oldest *supported* major version on which the bug reproduces.

Comment (nayuta-yanagisawa) [reply storage/spider/mysql-test/spider/bugfix/r/mdev_24523.result:15]: If the bug is reproducible on 10.2, please set the base branch to 10.2. Adding `--disable-ps-protocol` is also appreciated. Thanks.

File: sql/item_jsonfunc.h
Comment (nayuta-yanagisawa) [reply sql/item_jsonfunc.h:528]: Possibly. There is a related MDEV issue: https://jira.mariadb.org/browse/MDEV-24541

File: storage/spider/mysql-test/spider/bugfix/r/mdev_24523.result
Comment (yongxin-xu) [reply storage/spider/mysql-test/spider/bugfix/r/mdev_24523.result:15]: No problem, I'm taking a day off today, I will work on it tomorrow.

**Comment by yongxin-xu** [comment]
@nayuta-yanagisawa 
The bug is not reproducible on 10.2, where Spider would translate 
```sql
UPDATE t1 SET j = JSON_REPLACE(j, '$.a', 10, '$.c', '[1, 2]');
```
to
```sql
update `t1` set `j` = '{\"a\": 10, \"b\": [2, 3]}' where `i` = 1 and `j` = '{ \"a\": 1, \"b\": [2, 3]}' limit 1
```
on remote DB.

However, the [MDEV-24541](https://jira.mariadb.org/browse/MDEV-24541) is reproducible on 10.2.
I think I can create two pull requests, where 
1. the first pr fixes the `json_update` typo on 10.2 and adds a test case for MDEV-24541.
2. the second pr fixes the `json_replace` function error on 10.3 and adds a Spider test case for MDEV-24523. (This pr)

**Comment by nayuta-yanagisawa** [comment]
Thank you for your confirmation. The reason why the bug does not reproduce on 10.2 seems to be that 10.2 has no "direct update" functionality. In fact, https://github.com/MariaDB/server/commit/e53ef202bd7706b88760472472af5ae878065f4f is only merged to 10.3 or later.

**Comment by nayuta-yanagisawa** [comment]
I will soon review the patch. I think it's OK to merge the present patch first.

**Comment by yongxin-xu** [comment]
> I will soon review the patch. I think it's OK to merge the present patch first.

The current patch is based on branch 10.6, do I need to change the base branch to 10.3, add `disable-ps-protocol` and remove `log_output` in test files?

**Comment by nayuta-yanagisawa** [comment]
@yongxin-xu Ah, Yes, please. I forgot it.

**Comment by nayuta-yanagisawa** [comment]
Pushed to https://github.com/MariaDB/server/tree/bb-10.6-mdev-24523 to run full tests.

**Review by nayuta-yanagisawa — COMMENTED** [review]
Looks good to me. Just minor comments.

**Comment by nayuta-yanagisawa** [comment]
@vuvova Please take another look because the issue is in the server rather than in the Spider storage engine. 

The Spider SE reconstructs a query based on the return value of `func_name()` in some cases. Thus, if the return value is wrong, the Spider SE generates a wrong query.

File: storage/spider/mysql-test/spider/bugfix/include/mdev_24523_init.inc
Comment (yongxin-xu) [reply storage/spider/mysql-test/spider/bugfix/include/mdev_24523_init.inc]: Ok, these unused lines have been removed in the most recent commit. [4247e92](https://github.com/MariaDB/server/pull/1841/commits/4247e92bcae440ee73883af791782f1384d03cee)

Comment (nayuta-yanagisawa) [reply storage/spider/mysql-test/spider/bugfix/include/mdev_24523_init.inc]: I think you need to remove them from `mdev_24523_deinit.inc` too.

Comment (yongxin-xu) [reply storage/spider/mysql-test/spider/bugfix/include/mdev_24523_init.inc]: Ok, I pushed a new commit [ace36a3](https://github.com/MariaDB/server/pull/1841/commits/ace36a3afef996986ed8e9d8e28907f58b36e435) to remove that.

**Review by nayuta-yanagisawa — APPROVED** [review]
LGTM

**Comment by nayuta-yanagisawa** [comment]
Let's wait for the review by @vuvova.

**Comment by nayuta-yanagisawa** [comment]
I got a confirmation from @vuvova via private communication.  Will merge soon.

**Pull request merged**
